OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1467 @@
+            }
+        })
+        if (leftOverMethods.isEmpty) {
+            startSecondPhase(project)
+        } else Results(partialResults)
+    }
+
+    private[this] def continuationForTAC(m: Method)(eps: SomeEPS): PropertyComputationResult = {
+        eps match {
+            case UBP(tac: TACAI) if tac.tac.isDefined ⇒
+                processMethod(m)
+                leftOverMethods -= m
+                if (leftOverMethods.isEmpty && notStarted) {
+                    //if all methods are processed and the second phase has not started yet, the second phase is started.
+                    notStarted = false
+                    startSecondPhase(project)
+                } else NoResult
+            case _ ⇒
+                InterimPartialResult(
+                    Nil,
+                    Set(eps),
+                    continuationForTAC(m)
+                )
+        }
+    }
+
+    /**
+     * Collects information of intents, intent filters and callbacks from the given method m.
+     * @param m The processed method.
+     */
+    private def processMethod(m: Method): Unit = {
+        val body = m.body.get
+        val intentUseSites: mutable.Map[UByte, List[UByte]] = mutable.Map.empty[UByte, List[UByte]]
+        val tacCode = propertyStore.get(m, TACAI.key).get.ub.tac.get
+        body.collectWithIndex {
+            case PCAndInstruction(pc, in: INVOKESPECIAL) ⇒
+
+                val superInterfaceClasses = project.classHierarchy.allSupertypes(in.declaringClass.asObjectType, reflexive = true)
+                //find IntentFilters
+                if (superInterfaceClasses.contains(ObjectType("android/content/IntentFilter"))) {
+                    val stmts = tacCode.stmts
+                    val index = tacCode.pcToIndex(pc)
+                    val c = stmts(index)
+                    val useSites = tacCode.stmts(c.asNonVirtualMethodCall.receiver.asVar.definedBy.head).
+                        asAssignment.targetVar.usedBy.toChain.toList
+                    reconstructIntentfilter(c.asNonVirtualMethodCall, tacCode, useSites, m)
+                }
+
+                //find Intents
+                if (superInterfaceClasses.contains(intentOT)) {
+                    val stmts = tacCode.stmts
+                    val index = tacCode.pcToIndex(pc)
+                    val c = stmts(index).asNonVirtualMethodCall
+                    if (c.receiver.asVar.definedBy.head > -1) {
+                        val useSites = tacCode.stmts(c.receiver.asVar.definedBy.head).
+                            asAssignment.targetVar.usedBy.toChain.toList
+                        intentUseSites += index -> useSites
+                        reconstructIntent(c, tacCode, useSites, m)
+                    }
+                }
+        }
+
+        //find sending methods (virtual)
+        body.collect {
+            case in: Instruction if in.isInstanceOf[INVOKEVIRTUAL] &&
```

Comment:
Should be able to do `case in: INVOKEVIRTUAL` if senderClasses.contains(in.declaringClass) => in`. Also, I'm not sure why you don't move the check for the name here as well. You could probably do away with the `collect` here completely (that creates an intermediate data structure` and just do `foreach` (you can pattern match inside a foreach as well).

Reply:
By the way, why are you working on the raw bytecode instructions in the body at all? Could you do this using the TAC?